tests/unittests/Foundation/NSDecimalNumberTests.mm

```diff
@@ -0,0 +1,356 @@
+    EXPECT_ARRAY_EQUAL(mantissa, result._mantissa);
+    EXPECT_EQ(5, result._length);
+    EXPECT_EQ(0, result._isNegative);
+    EXPECT_EQ(-18, result._exponent);
+
+    //
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:10] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:0.5] autorelease]);
+
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:1000] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:-989.5] autorelease]);
+
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithDouble:-10.5] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:10] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:-20.5] autorelease]);
+
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:-1e3] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:1010.5] autorelease]);
+
+    NSDecimalNumber* d = [[[NSDecimalNumber alloc] initWithDouble:1012345678912345e-120] autorelease];
+    TEST_SUBTRACT(d, d, [NSDecimalNumber zero]);
+
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithInteger:2] autorelease], [NSDecimalNumber one], [NSDecimalNumber one]);
+
+    TEST_SUBTRACT([[[NSDecimalNumber alloc] initWithInt:(2 * USHRT_MAX)] autorelease],
+                  [[[NSDecimalNumber alloc] initWithInt:USHRT_MAX] autorelease],
+                  [[[NSDecimalNumber alloc] initWithInt:USHRT_MAX] autorelease]);
+}
+
+TEST(NSDecimalNumber, addition) {
+    // Note: Subtraction tests addition, since NSDecimalSubtract gets converted to NSDecimalAdd
+    unsigned short mantissa[NSDecimalMaxSize] = { 5272, 10, 0, 0, 0, 0, 0, 0 };
+    NSDecimalNumber* dNumber = [[[NSDecimalNumber alloc] initWithLongLong:6560451.0] autorelease];
+    NSDecimal decimal1 = [dNumber decimalValue];
+
+    dNumber = [[[NSDecimalNumber alloc] initWithDouble:45869] autorelease];
+    double maxx = DBL_MAX;
+    NSDecimal decimal2 = [dNumber decimalValue];
+
+    NSDecimal result = {};
+
+    NSDecimalAdd(&result, &decimal1, &decimal2, NSRoundPlain);
+
+    EXPECT_EQ(2, result._length);
+    EXPECT_EQ(1, result._exponent);
+    EXPECT_EQ(0, result._isNegative);
+
+    EXPECT_ARRAY_EQUAL(mantissa, result._mantissa);
+
+    TEST_ADDITION([[[NSDecimalNumber alloc] initWithDouble:19605] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:6536] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:26141] autorelease]);
+
+    TEST_ADDITION([[[NSDecimalNumber alloc] initWithDouble:10.5] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:-1e3] autorelease],
+                  [[[NSDecimalNumber alloc] initWithDouble:-989.5] autorelease]);
```

Comment:
Can you add a few tests where the exponents differ? Especially interesting would be if they differed enough to incur a loss of precision (though I don't think that specific feature works currently, it's worth writing for now as a DISABLED_TEST)